Frontend/default.nix

```diff
@@ -1,28 +1,213 @@
-    in {
+
+  perSystem = { config, self', lib, system, ... }:
+    let
+      nodejs = pkgs.nodejs-14_x;
+
+      pkgs = import inputs.nixpkgs {
+        inherit system;
+        overlays = [
+          inputs.purifix.overlay
+        ];
+      };
+
+      localPackages = pkgs.purifix {
+        src = ./.;
+      };
+
+      nodePackages = inputs.dream2nix.lib.makeFlakeOutputs {
+        systems = [ system ];
+        source = ./node;
+        projects = {
+          "atlas-ui" = {
+            name = "atlas-ui";
+            subsystem = "nodejs";
+            translator = "package-lock";
+            builder = "strict-builder";
+            subsystemInfo = {
+              nodejs = 14;
+            };
+          };
+        };
+      };
+
+      nodeDependencies = nodePackages.packages.${system}.atlas-ui.lib;
+
+      typescript-language-server = pkgs.symlinkJoin {
+        name = "typescript-language-server";
+        nativeBuildInputs = [ pkgs.makeWrapper ];
+        paths = [
+          pkgs.nodePackages.typescript
+          pkgs.nodePackages.typescript-language-server
+        ];
+        postBuild = ''
+          wrapProgram $out/bin/typescript-language-server \
+            --add-flags "--tsserver-path $out/bin/tsserver"
+        '';
+      };
+
+      webpack-dev-server = pkgs.writeShellApplication {
+        name = "webpack-dev-server-wrapped";
+        runtimeInputs = [ nodejs ];
+        text = ''
+          export NODE_PATH="${nodeDependencies}/node_modules"
+          PIPE="$1"
+          shift
+          tail -n1 -f "$PIPE" | node ${./node/watch.js} "$@"
+        '';
+      };
+
+      purifix-watch =
+        pkgs.writeShellApplication {
+          name = "purifix-watch";
+          runtimeInputs = [
+            pkgs.watchexec
+          ] ++ localPackages.beckn-common.develop.buildInputs;
+          text = ''
+            PIPE=$1
+            shift
+            SRC_FOLDERS=()
+            SRC_GLOBS=()
+            for SRC_FOLDER in "$@"; do
+              if [ -d "$SRC_FOLDER/src" ]; then
+                SRC_FOLDERS+=( -w "$SRC_FOLDER/src" )
+                SRC_GLOBS+=( "\"$SRC_FOLDER/src/**/*.purs\"" )
+              fi
+            done
+            watchexec "''${SRC_FOLDERS[@]}" -r -f "**/*.purs" -- \
+              "echo 'suspend' >$PIPE;purifix ''${SRC_GLOBS[*]} && echo 'ready' >$PIPE"
+          '';
+        };
+
+      start-app-devserver = { target, env, platform }: pkgs.writeShellApplication rec {
+        name = "watch-${target}-${platform}-${env}";
+        text = ''
+          FLAKE_ROOT=''$(${lib.getExe config.flake-root.package})
+          cd "$FLAKE_ROOT"
+          export ENV=${env}
+          export TARGET=./Frontend/${target}
+          export CONFIG_FILE
+          CONFIG_FILE=$(realpath "$TARGET/webpack.${platform}.js")
+
+          export PIPE=/tmp/${name}
+          mkfifo $PIPE
+          trap 'rm -f $PIPE' EXIT
+
+          ${lib.getExe config.proc.groups.frontend-dev.package}
+        '';
+      };
```

Comment:
I don't think we need this `start-app-devserver` indirection. It is used in two places, and can be inlined in proc-flake definition. Also `$PIPE` should be stored under project directory (eg: `./Frontend/tmp/${name}`). And proc-flake already cd's to FLAKE_ROOT.

This definition is much more straightforward to read as a new person in the team:

```nix
      proc.groups =
        let
          mkWatchProcesses = { target, platform, env }: {
            webpack-watch.command = ''
              ${lib.getExe webpack-dev-server} \
                ./tmp/pipes/${target} \
                --env ${env} \
                --config $(realpath "./Frontend/${target}/webpack.${platform}.js")
            '';
            purifix-watch.command = ''
              cd ./Frontend/${target}
              ${lib.getExe purifix-watch} ./tmp/pipes/${target} . ../ui-common
            '';
          };
        in
        {
          ui-customer-start.processes = mkWatchProcesses {
            target = "ui-customer";
            platform = "android";
            env = "master";
          };
          ui-driver-start.processes = mkWatchProcesses {
            target = "ui-driver";
            platform = "android";
            env = "master";
          };
        };
```

Reply:
Also, let's move all the watch related stuff to its own file `watch.nix` or something.

Reply:
Question: what happens if the user tries to start the watch twice in different terminals at the same time? Can the pipe be reused as a lock to prevent that?